blech32/blech32.go

```diff
@@ -0,0 +1,247 @@
+}
+
+// ConvertBits converts a byte slice where each byte is encoding fromBits bits,
+// to a byte slice where each byte is encoding toBits bits.
+func ConvertBits(data []byte, fromBits, toBits uint8, pad bool) ([]byte, error) {
+	if fromBits < 1 || fromBits > 8 || toBits < 1 || toBits > 8 {
+		return nil, fmt.Errorf("only bit groups between 1 and 8 allowed")
+	}
+
+	// The final bytes, each byte encoding toBits bits.
+	var regrouped []byte
+
+	// Keep track of the next byte we create and how many bits we have
+	// added to it out of the toBits goal.
+	nextByte := byte(0)
+	filledBits := uint8(0)
+
+	for _, b := range data {
+
+		// Discard unused bits.
+		b = b << (8 - fromBits)
+
+		// How many bits remaining to extract from the input data.
+		remFromBits := fromBits
+		for remFromBits > 0 {
+			// How many bits remaining to be added to the next byte.
+			remToBits := toBits - filledBits
+
+			// The number of bytes to next extract is the minimum of
+			// remFromBits and remToBits.
+			toExtract := remFromBits
+			if remToBits < toExtract {
+				toExtract = remToBits
+			}
+
+			// Add the next bits to nextByte, shifting the already
+			// added bits to the left.
+			nextByte = (nextByte << toExtract) | (b >> (8 - toExtract))
+
+			// Discard the bits we just extracted and get ready for
+			// next iteration.
+			b = b << toExtract
+			remFromBits -= toExtract
+			filledBits += toExtract
+
+			// If the nextByte is completely filled, we add it to
+			// our regrouped bytes and start on the next byte.
+			if filledBits == toBits {
+				regrouped = append(regrouped, nextByte)
+				filledBits = 0
+				nextByte = 0
+			}
+		}
+	}
+
+	// We pad any unfinished group if specified.
+	if pad && filledBits > 0 {
+		nextByte = nextByte << (toBits - filledBits)
+		regrouped = append(regrouped, nextByte)
+		filledBits = 0
+		nextByte = 0
+	}
+
+	// Any incomplete group must be <= 4 bits, and all zeroes.
+	if filledBits > 0 && (filledBits > 4 || nextByte != 0) {
+		return nil, fmt.Errorf("invalid incomplete group")
+	}
+
+	return regrouped, nil
+}
+
+// For more details on the checksum calculation, please refer to BIP 173.
+func blech32Checksum(hrp string, data []byte) []byte {
+	// Convert the bytes to list of integers, as this is needed for the
+	// checksum calculation.
+	integers := make([]int, len(data))
+	for i, b := range data {
+		integers[i] = int(b)
+	}
+	values := append(blech32HrpExpand(hrp), integers...)
+	values = append(values, []int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0}...) //6->12 compared to bech32
+	polymod := blech32Polymod(values) ^ 1
+	var res []byte
+	for i := 0; i < 12; i++ { //6 -> 12 compared to bech32
+		res = append(res, byte((polymod>>uint(5*(11-i)))&31)) //5 -> 11 compared to bech32
+	}
+	return res
+}
+
+// For more details on the polymod calculation, please refer to BIP 173.
+func blech32Polymod(values []int) int {
+	chk := 1
+	for _, v := range values {
+		b := chk >> 50                      //25->55 compared to bech32
```

Comment:
Is this correct? In the comment it's said `55` but we're shifting `50`. Is this intended?

Reply:
Fromm your [comment in the issue](https://github.com/vulpemventures/go-elements/issues/41#issuecomment-632727192),  check out the [2nd difference](https://github.com/Simplexum/python-elementstx/blob/065363ea6f469c8f860422623151802b06fba391/elementstx/liquid_addr.py#L36) you pointed.

Reply:
@sekulicd

Reply:
You are right after fixing this test case is passing, can't believe I missed this after so much double-checks :)